File: src/audio/tests/processor_test.cpp
/***

  Oak Video Editor - Non-Linear Video Editor
  Copyright (C) 2026 Oak Team

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include <cmath>
#include <vector>

#include <gtest/gtest.h>

#include "audio/manager.h"
#include "audio/processor.h"

namespace
{

constexpr int kSampleFmtF32P = 4; // olive::core::SampleFormat::f32_p
constexpr uint64_t kLayoutStereo = 0x3;

struct ProcessorHandle {
	OakAudioProcessor h = oakaudio_processor_init();
	~ProcessorHandle() { oakaudio_processor_free(&h); }
};

// Feed a full buffer through the processor and return the total number of
// output frames produced (input drained + flushed).
int convert_all(OakAudioProcessor p, const std::vector<std::vector<float>> &in,
				int chunk)
{
	const int channels = int(in.size());
	const size_t nch = size_t(channels);
	std::vector<const float *> in_planes(nch);
	std::vector<std::vector<float>> out_store(nch);
	std::vector<float *> out_planes(nch);
	for (int ch = 0; ch < channels; ch++) {
		in_planes[size_t(ch)] = in[size_t(ch)].data();
		out_store[size_t(ch)].resize(size_t(chunk) * 4 + 4096);
		out_planes[size_t(ch)] = out_store[size_t(ch)].data();
	}

	int total = 0;
	const int frames = int(in[0].size());
	for (int pos = 0; pos < frames; pos += chunk) {
		const int n = std::min(chunk, frames - pos);
		std::vector<const float *> window(nch);
		for (int ch = 0; ch < channels; ch++) {
			window[size_t(ch)] = in[size_t(ch)].data() + pos;
		}
		const int produced = oakaudio_processor_convert(
			p, window.data(), n, out_planes.data(), int(out_store[0].size()));
		if (produced < 0) {
			return produced;
		}
		total += produced;
	}

	EXPECT_EQ(oakaudio_processor_flush(p), OAKAUDIO_OK);
	// Drain the resampler's internal delay
	for (int guard = 0; guard < 64; guard++) {
		const int produced = oakaudio_processor_convert(
			p, nullptr, 0, out_planes.data(), int(out_store[0].size()));
		if (produced <= 0) {
			break;
		}
		total += produced;
	}
	return total;
}

std::vector<std::vector<float>> make_sine(int channels, int frames, int rate)
{
	const size_t nch = size_t(channels);
	std::vector<std::vector<float>> data(nch);
	for (int ch = 0; ch < channels; ch++) {
		data[size_t(ch)].resize(size_t(frames));
		for (int i = 0; i < frames; i++) {
			data[size_t(ch)][size_t(i)] =
				0.5f * std::sin(2.0 * M_PI * 440.0 * i / rate);
		}
	}
	return data;
}

} // namespace

TEST(OakAudioProcessor, InitFree)
{
	const int before = oakaudio_debug_alive_count();
	{
		ProcessorHandle p;
		ASSERT_NE(p.h.ctx, nullptr);
		EXPECT_EQ(oakaudio_debug_alive_count(), before + 1);
	}
	EXPECT_EQ(oakaudio_debug_alive_count(), before);

	// free is a no-op on NULL / empty handles
	oakaudio_processor_free(nullptr);
	OakAudioProcessor empty = {};
	oakaudio_processor_free(&empty);
}

TEST(OakAudioProcessor, OpenCloseIsOpen)
{
	ProcessorHandle p;
	EXPECT_EQ(oakaudio_processor_is_open(p.h), 0);

	EXPECT_EQ(oakaudio_processor_open(p.h, 44100, kLayoutStereo, kSampleFmtF32P,
									48000, kLayoutStereo, kSampleFmtF32P, 1.0),
			  OAKAUDIO_OK);
	EXPECT_EQ(oakaudio_processor_is_open(p.h), 1);

	// Error path: opening an open processor
	EXPECT_EQ(oakaudio_processor_open(p.h, 44100, kLayoutStereo, kSampleFmtF32P,
									48000, kLayoutStereo, kSampleFmtF32P, 1.0),
			  OAKAUDIO_E_STATE);

	EXPECT_EQ(oakaudio_processor_close(p.h), OAKAUDIO_OK);
	EXPECT_EQ(oakaudio_processor_is_open(p.h), 0);
}

TEST(OakAudioProcessor, OpenInvalidArgs)
{
	ProcessorHandle p;
	// Unsupported output format (only f32p is delivered)
	EXPECT_EQ(oakaudio_processor_open(p.h, 44100, kLayoutStereo, kSampleFmtF32P,
									48000, kLayoutStereo, 1, 1.0),
			  OAKAUDIO_E_INVALID);
	// Bad sample rate
	EXPECT_EQ(oakaudio_processor_open(p.h, 0, kLayoutStereo, kSampleFmtF32P,
									48000, kLayoutStereo, kSampleFmtF32P, 1.0),
			  OAKAUDIO_E_INVALID);
	EXPECT_EQ(oakaudio_processor_is_open(p.h), 0);

	// Empty handle
	OakAudioProcessor empty = {};
	EXPECT_EQ(oakaudio_processor_open(empty, 44100, kLayoutStereo,
									kSampleFmtF32P, 48000, kLayoutStereo,
									kSampleFmtF32P, 1.0),
			  OAKAUDIO_E_INVALID);
	EXPECT_EQ(oakaudio_processor_is_open(empty), OAKAUDIO_E_INVALID);
	EXPECT_EQ(oakaudio_processor_close(empty), OAKAUDIO_E_INVALID);
	EXPECT_EQ(oakaudio_processor_flush(empty), OAKAUDIO_E_INVALID);
}

TEST(OakAudioProcessor, ConvertResample441To48)
{
	const int before = oakaudio_debug_alive_count();
	ProcessorHandle p;
	ASSERT_EQ(oakaudio_processor_open(p.h, 44100, kLayoutStereo,
									kSampleFmtF32P, 48000, kLayoutStereo,
									kSampleFmtF32P, 1.0),
			  OAKAUDIO_OK);

	const int in_frames = 44100; // one second
	const auto sine = make_sine(2, in_frames, 44100);
	const int produced = convert_all(p.h, sine, 4096);
	ASSERT_GE(produced, 0);

	// One second at 44.1k must become (within resampler tolerance) one
	// second at 48k.
	EXPECT_NEAR(produced, 48000, 200);

	// Re-open check for leaks
	oakaudio_processor_close(p.h);
	oakaudio_processor_free(&p.h);
	EXPECT_EQ(oakaudio_debug_alive_count(), before);
}

TEST(OakAudioProcessor, ConvertSilenceStaysSilent)
{
	ProcessorHandle p;
	ASSERT_EQ(oakaudio_processor_open(p.h, 48000, kLayoutStereo,
									kSampleFmtF32P, 48000, kLayoutStereo,
									kSampleFmtF32P, 1.0),
			  OAKAUDIO_OK);

	std::vector<std::vector<float>> silence(2, std::vector<float>(4096, 0.0f));
	std::vector<std::vector<float>> out(2, std::vector<float>(8192, -1.0f));
	std::vector<const float *> in_planes = { silence[0].data(),
										 silence[1].data() };
	std::vector<float *> out_planes = { out[0].data(), out[1].data() };

	const int produced = oakaudio_processor_convert(
		p.h, in_planes.data(), 4096, out_planes.data(), 8192);
	ASSERT_GT(produced, 0);
	EXPECT_EQ(produced, 4096); // same rate in/out: 1:1 frames
	for (int i = 0; i < produced; i++) {
		EXPECT_FLOAT_EQ(out[0][size_t(i)], 0.0f);
		EXPECT_FLOAT_EQ(out[1][size_t(i)], 0.0f);
	}
}

TEST(OakAudioProcessor, ConvertTempo)
{
	ProcessorHandle p;
	ASSERT_EQ(oakaudio_processor_open(p.h, 48000, kLayoutStereo,
									kSampleFmtF32P, 48000, kLayoutStereo,
									kSampleFmtF32P, 1.5),
			  OAKAUDIO_OK);

	const auto sine = make_sine(2, 48000, 48000);
	const int produced = convert_all(p.h, sine, 4096);
	ASSERT_GE(produced, 0);

	// 1.5x tempo: one second of input becomes roughly 2/3 second of
	// output (atempo works on correlated windows, so allow slack)
	EXPECT_NEAR(produced, int(48000 / 1.5), 3000);
}

TEST(OakAudioProcessor, ConvertErrorPaths)
{
	ProcessorHandle p;

	// Convert on a closed processor
	float dummy = 0.0f;
	float *out_planes[1] = { &dummy };
	const float *in_planes[1] = { &dummy };
	EXPECT_EQ(oakaudio_processor_convert(p.h, in_planes, 1, out_planes, 1),
			  OAKAUDIO_E_STATE);
	EXPECT_EQ(oakaudio_processor_flush(p.h), OAKAUDIO_E_STATE);

	// Empty handle
	OakAudioProcessor empty = {};
	EXPECT_EQ(oakaudio_processor_convert(empty, in_planes, 1, out_planes, 1),
			  OAKAUDIO_E_INVALID);

	// NULL input planes with frames
	ASSERT_EQ(oakaudio_processor_open(p.h, 48000, kLayoutStereo,
									kSampleFmtF32P, 48000, kLayoutStereo,
									kSampleFmtF32P, 1.0),
			  OAKAUDIO_OK);
	EXPECT_EQ(oakaudio_processor_convert(p.h, nullptr, 10, out_planes, 1),
			  OAKAUDIO_E_INVALID);
	// Negative counts
	EXPECT_EQ(oakaudio_processor_convert(p.h, in_planes, -1, out_planes, 1),
			  OAKAUDIO_E_INVALID);

	EXPECT_EQ(oakaudio_processor_flush(p.h), OAKAUDIO_OK);
}

File: crates/oakaudio/tests/processor_test.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! AudioProcessor contract tests (processor.rs), calling the public Rust
//! API. The conversion runs a real FFmpeg filter graph
//! (aresample/aformat/atempo), so resampling and time-stretch have filter
//! latency: the exact frame counts are drained after `flush`, while
//! identity conversion is an immediate passthrough.

use oakaudio::error::{Error, OAKAUDIO_E_INVALID, OAKAUDIO_E_STATE, OAKAUDIO_OK};
use oakaudio::params::{AudioParams, SampleFormat};
use oakaudio::processor::Processor;

/// Stereo f32_p planes of `frames` ramp samples.
fn ramp_planes(frames: usize) -> Vec<Vec<f32>> {
	vec![
		(0..frames).map(|i| i as f32 * 0.01).collect(),
		(0..frames).map(|i| -(i as f32) * 0.01).collect(),
	]
}

/// Stereo planar-f32 params at `rate`.
fn params(rate: i32) -> AudioParams {
	AudioParams {
		sample_rate: rate,
		channel_layout: 0x3,
		format: SampleFormat::F32Planar,
	}
}

fn open_identity(p: &Processor) -> Result<(), Box<dyn std::error::Error>> {
	p.open(params(48000), params(48000), 1.0)
}

/// Pointer arrays for the convert call.
fn plane_ptrs(planes: &[Vec<f32>]) -> Vec<*const f32> {
	planes.iter().map(|p| p.as_ptr()).collect()
}

fn plane_mut_ptrs(planes: &mut [Vec<f32>]) -> Vec<*mut f32> {
	planes.iter_mut().map(|p| p.as_mut_ptr()).collect()
}

/// The crate error code of a failed call (the API surfaces
/// `Box<dyn std::error::Error>`).
fn code(err: Box<dyn std::error::Error>) -> i32 {
	err.downcast_ref::<oakaudio::error::Error>()
		.map(|e| e.code())
		.unwrap_or(-1)
}

/// init yields a closed processor; is_open is false before open and true
/// after; close returns it to closed without error.
#[test]
fn processor_open_isopen_close() {
	let p = Processor::init();
	assert_eq!(p.is_open().unwrap(), false);
	open_identity(&p).unwrap();
	assert_eq!(p.is_open().unwrap(), true);
	p.close().unwrap();
	assert_eq!(p.is_open().unwrap(), false);
}

/// open with matching in/out rate and format is an identity passthrough:
/// convert returns the same frame count and samples within 1e-6.
#[test]
fn identity_convert_passthrough() {
	let p = Processor::init();
	open_identity(&p).unwrap();

	let planes = ramp_planes(32);
	let in_ptrs = plane_ptrs(&planes);
	let mut out = vec![vec![0f32; 32]; 2];
	let mut out_ptrs = plane_mut_ptrs(&mut out);

	let n = p.convert(in_ptrs.as_ptr(), 32, out_ptrs.as_ptr(), 32).unwrap();
	assert_eq!(n, 32);
	for ch in 0..2 {
		for i in 0..32 {
			assert!(
				(out[ch][i] - planes[ch][i]).abs() < 1e-6,
				"ch{ch}[{i}]: {} vs {}",
				out[ch][i],
				planes[ch][i]
			);
		}
	}
}

/// convert with an output capacity smaller than the produced frames returns
/// the produced count clamped to capacity and fills up to capacity.
#[test]
fn convert_capacity_truncation() {
	let p = Processor::init();
	open_identity(&p).unwrap();

	let planes = ramp_planes(32);
	let in_ptrs = plane_ptrs(&planes);
	let mut out = vec![vec![9.9f32; 10]; 2];
	let mut out_ptrs = plane_mut_ptrs(&mut out);

	let n = p.convert(in_ptrs.as_ptr(), 32, out_ptrs.as_ptr(), 10).unwrap();
	assert_eq!(n, 10);
	for ch in 0..2 {
		for i in 0..10 {
			assert_eq!(out[ch][i], planes[ch][i]);
		}
	}

	// The graph has already drained; nothing further to pull.
	let mut out2 = vec![vec![0f32; 32]; 2];
	let mut out2_ptrs = plane_mut_ptrs(&mut out2);
	let n = p.convert(in_ptrs.as_ptr(), 0, out2_ptrs.as_ptr(), 32).unwrap();
	assert_eq!(n, 0);
}

/// open with a zero/negative rate or a wrong output format is rejected and
/// leaves the processor closed; convert before open is a state error.
#[test]
fn open_invalid_params() {
	let p = Processor::init();
	assert_eq!(
		code(p.open(params(0), params(48000), 1.0).unwrap_err()),
		OAKAUDIO_E_INVALID
	);
	assert_eq!(p.is_open().unwrap(), false);
	// The output format is forced to planar f32 by the processor contract.
	let mut wrong_out = params(48000);
	wrong_out.format = SampleFormat::F32;
	assert_eq!(
		code(p.open(params(48000), wrong_out, 1.0).unwrap_err()),
		OAKAUDIO_E_INVALID
	);
	assert_eq!(p.is_open().unwrap(), false);

	// convert before open is a state error; a non-positive speed is
	// rejected on open.
	assert_eq!(
		code(p.convert(std::ptr::null(), 0, std::ptr::null(), 0).unwrap_err()),
		OAKAUDIO_E_STATE
	);
	assert_eq!(
		code(p.open(params(48000), params(48000), 0.0).unwrap_err()),
		OAKAUDIO_E_INVALID
	);
}

/// Resampling to half rate halves the frame count (44100 -> 22050). The
/// real resampler holds samples back (filter delay), so the frames are
/// drained after `flush`; flush then keeps the processor open.
#[test]
fn resample_and_flush() {
	let p = Processor::init();
	p.open(params(44100), params(22050), 1.0).unwrap();

	// 1 second of input keeps the resampler delay well below the signal.
	let frames = 44100;
	let planes = ramp_planes(frames);
	let in_ptrs = plane_ptrs(&planes);
	let mut out = vec![vec![0f32; frames]; 2];
	let mut out_ptrs = plane_mut_ptrs(&mut out);

	let mut total = p
		.convert(in_ptrs.as_ptr(), frames as i32, out_ptrs.as_ptr(), frames as i32)
		.unwrap();
	p.flush().unwrap();
	// Drain the resampler delay after end-of-input.
	while total < frames as i32 {
		let n = p.convert(std::ptr::null(), 0, out_ptrs.as_ptr(), frames as i32).unwrap();
		if n == 0 {
			break;
		}
		total += n;
	}
	assert!(
		(total - 22050).abs() <= 2,
		"half-rate output must halve the frame count (got {total})"
	);
	assert_eq!(p.is_open().unwrap(), true);
}

/// A tempo factor != 1.0 time-stretches: tempo 2.0 halves the frame count
/// (drained after `flush`; atempo needs a full analysis window before it
/// produces output) and the processor stays open.
#[test]
fn tempo_stretch() {
	let p = Processor::init();
	p.open(params(48000), params(48000), 2.0).unwrap();

	// 1 second of input: many atempo windows (1024 samples at 48 kHz).
	let frames = 48000;
	let planes = ramp_planes(frames);
	let in_ptrs = plane_ptrs(&planes);
	let mut out = vec![vec![0f32; frames]; 2];
	let mut out_ptrs = plane_mut_ptrs(&mut out);

	let mut total = p
		.convert(in_ptrs.as_ptr(), frames as i32, out_ptrs.as_ptr(), frames as i32)
		.unwrap();
	p.flush().unwrap();
	while total < frames as i32 {
		let n = p.convert(std::ptr::null(), 0, out_ptrs.as_ptr(), frames as i32).unwrap();
		if n == 0 {
			break;
		}
		total += n;
	}
	assert!(
		(total - 24000).abs() <= 2400,
		"tempo 2.0 must halve the frame count (got {total})"
	);
	assert_eq!(p.is_open().unwrap(), true);
	p.close().unwrap();

	// Re-opening a closed processor works; opening an open one is a state
	// error.
	open_identity(&p).unwrap();
	assert_eq!(code(open_identity(&p).unwrap_err()), OAKAUDIO_E_STATE);
}

/// The error mapping is intact.
#[test]
fn error_codes() {
	assert_eq!(Error::Invalid.code(), OAKAUDIO_E_INVALID);
	assert_eq!(Error::State.code(), OAKAUDIO_E_STATE);
	assert_eq!(Error::Failed("x".to_string()).code(), -60003);
	assert_eq!(oakaudio::error::OAKAUDIO_OK, OAKAUDIO_OK);
}
